resources/design-document.md

```diff
@@ -0,0 +1,217 @@
+recipe with adjustable requirements according to number of desired servings, and with the ability to toggle
+between systems of measurement.
+
+Current expected endpoints(`CreateRecipe`, `GetRecipe`, `GetAllRecipes`, `UpdateRecipe`, `DeleteRecipe`,  `CreateList`, `GetList`, `UpdateList`)
+
+Recipes and shopping lists will be stored in their respective DynamoDB tables. The homepage will display the 
+names of all recipes along with the option to edit, delete, navigate to a details page, or add the recipe to 
+the shopping list. The homepage will also have the option to create and save a new recipe to the table, and 
+navigate to the shopping list page.
+
+## 6. API
+
+---
+
+### 6.1 Public Models
+
+#### RecipeModel
+
+```
+String userId
+String recipeName
+Integer servings
+Map<Integer, String> recipeSteps
+List<Ingredient> ingredients
+Integer calories
+Enum systemOfMeasurement
+    - METRIC
+    - IMPERIAL
+```
+
+#### IngredientModel
+
+```
+String ingredientName
+Double amount
+Enum unitOfMeasurement
+    - OUNCES
+    - FLUID_OUNCES
+    - CUPS
+    - TEASPOONS
+    - TABLESPOONS
+    - GRAMS
+    - MILLIGRAMS
+    - MILLILITERS
+    - COUNT
+```
+
+#### ListModel
+
+```
+String userId
+Map<Ingredient, Integer(quantity)>
+```
+
+
+
+### 6.2 CreateRecipe Endpoint
+
+---
+
+- Accepts `POST` requests to `/recipes`
+- Accepts data to create a new recipe with a provided name, a given customerId, a number of servings,calories, a list of steps for the recipe, and a system of measurement
+
+### 6.3 GetRecipe Endpoint
+
+---
+
+- Accepts `GET` requests to `/recipes/:recipeName`
+- Accepts a user ID and a recipe name and returns the corresponding RecipeModel
+    - if the recipe does not exist, it throws a `RecipeNotFoundException`
+
+### 6.4 GetAllRecipes Endpoint
+
+---
+
+- Accepts `GET` requests to `/recipes`
+- Accepts a userId and returns all recipes from table with matching id
+
+### 6.5 UpdateRecipe Endpoint
+
+---
+
+- Accepts `PUT` requests to `/recipes/:recipeName`
+- Accepts data to update a recipe such as new ingredients, changing their measurements, or removing them
+- if the recipe does not exist, it throws a `RecipeNotFoundException`
+
+### 6.6 DeleteRecipe Endpoint
+
+---
+
+- Accepts `DELETE` requests to `/recipes/:recipeName`
+- Accepts a userId and recipeName to remove a recipe from the table
+  - if the recipe does not exist, it throws a `RecipeNotFoundException`
+
+### 6.7 CreateList Endpoint
+
+---
+
+- Accepts `POST` requests to `/list`
+- Accepts a userId and creates a single list within the table
+- Each user should only have one list
+
+### 6.8 GetList Endpoint
+
+---
+
+- Accepts `GET` requests to `/list`
+- Accepts a userId and retrieves the single list within the table
+
+### 6.9 UpdateList Endpoint
+
+---
+
+- Accepts `PUT` requests to `/list`
+- Accepts new lists of ingredients and adds them to existing list
+- User will have the option to edit items from list or clear list
+
+## 7. Tables
```

Comment:
Have you thought about how/where you could include a GSI in the design?

Reply:
Yes, I forgot we needed one of those. I was thinking I could move the "Filtering recipes through a range of calories" story from out of scope, and query recipes within a range of calories, still using the 'userId' as the hashkey, but using 'calories' as the rangekey instead of 'recipeName'.